### DISCUSSION TITLE
How to add my own methods to `Guard`?

### DISCUSSION THREAD

**Discussion by Kilazur** [CommunityToolkit/dotnet#975] [Q&A] [answered]
I'd like to create custom guard methods but be able to use them transparently as if they were native from the original NuGet package.

Example:

```
void MyMethod(string[] values)
{
    Guard.IsNotNullOrEmpty(values);
    Guard.AllValid(values);
}

// ???
public static void IsNotNullOrEmpty(this Guard guard, string[] values, [CallerMemberName] name = "")
{
    Guard.IsNotNull(values);
    Guard.IsNotEmpty(values);
}

public static void AllValid(this Guard guard, string[] values, [CallerMemberName] name = "")
{
    if (values.Any(v => v.Length < 4)
    {
        throw new ArgumentException...
    }
}
```

But `Guard` being a static type, it cannot be extended. And you cannot taker advantage of `partial` from types in another assembly.

Am I doomed to not do this?

**Comment by hawkerm** [comment, marked as answer]
You'd just have to create your own static class for now.

Eventually C# will get some extension feature that allows you to add on to a static class, but that's been coming soon for a while now. (Sometimes referred to as "Extension everything" in the past, but then got superseded by "roles", I believe that discussion had happened here: https://github.com/dotnet/csharplang/issues/5485)

They're doing some restructuring on the C# repo, so I don't even know where the latest info tracking its status is, this was the best I could find as a pointer: https://github.com/dotnet/csharplang/issues/8431
